spec/lib/diaspora/federated/generator_spec.rb

```diff
@@ -0,0 +1,17 @@
+require "spec_helper"
+
+describe "comment on reshare" do
+  before do
+    @status_message = FactoryGirl.build(:status_message, author: alice.person, public: true)
+    @reshare = FactoryGirl.build(:reshare, root: @status_message, author: bob.person)
+    @comment = FactoryGirl.build(:comment, post: @reshare, text: "reshare comment")
+  end
+
+  it "delivers comment also to root author" do
+    # opts = {:additional_subscribers => [alice.person]}
+    allow_any_instance_of(Federated::Generator).to receive(:create!)
+    # expect(Postzord::Dispatcher).to receive(:defer_build_and_post).with(bob, @comment, opts)
+    expect(Postzord::Dispatcher).to receive(:defer_build_and_post)
```

Comment:
Let's trace the call chain. Creating a public comment via the UI:

* https://github.com/diaspora/diaspora/blob/develop/app/controllers/comments_controller.rb#L15
* https://github.com/diaspora/diaspora/blob/develop/app/services/comment_service.rb#L15
* https://github.com/diaspora/diaspora/blob/develop/app/models/user/social_actions.rb#L4
* https://github.com/diaspora/diaspora/blob/develop/app/models/comment.rb#L94
* https://github.com/diaspora/diaspora/blob/develop/lib/federated/generator.rb#L11
*  https://github.com/diaspora/diaspora/blob/develop/lib/postzord/dispatcher.rb#L50
* https://github.com/diaspora/diaspora/blob/develop/app/workers/deferred_dispatch.rb#L16

So no, `save!` doesn't trigger a callback that calls `create!`, rather it's called explicitly and is what calls `save!`. Let's pause here for a second and look at ...

Receiving a public comment via federation:

* https://github.com/diaspora/diaspora/blob/develop/app/controllers/publics_controller.rb#L24
* https://github.com/diaspora/diaspora/blob/develop/app/workers/receive_unencrypted_salmon.rb#L12
* https://github.com/diaspora/diaspora/blob/develop/lib/postzord/receiver.rb#L14
* https://github.com/diaspora/diaspora/blob/develop/lib/postzord/receiver/public.rb#L24
* https://github.com/diaspora/diaspora/blob/develop/lib/postzord/receiver/public.rb#L54
* https://github.com/diaspora/diaspora/blob/develop/lib/diaspora/parser.rb#L13
* Effectively `Comment.from_xml`, `from_xml` is ROXML magic (the `xml_attr` stuff in the models). No callbacks.
* https://github.com/diaspora/diaspora/blob/develop/lib/postzord/receiver/public.rb#L59
* https://github.com/diaspora/diaspora/blob/develop/lib/postzord/receiver/public.rb#L67 (We're persisted to the DB now)
* https://github.com/diaspora/diaspora/blob/develop/lib/postzord/receiver/public.rb#L32-L33 (https://github.com/diaspora/diaspora/blob/develop/app/models/comment.rb#L10, https://github.com/diaspora/diaspora/blob/develop/lib/diaspora/relayable.rb#L43)
* https://github.com/diaspora/diaspora/blob/develop/lib/postzord/receiver/public.rb#L45
* https://github.com/diaspora/diaspora/blob/develop/lib/diaspora/relayable.rb#L89

Now we're at `Postzord::Dispatcher.build(...).post` again, so the code paths converge here.

* https://github.com/diaspora/diaspora/blob/develop/lib/postzord/dispatcher.rb#L34
* https://github.com/diaspora/diaspora/blob/develop/lib/postzord/dispatcher.rb#L66
* https://github.com/diaspora/diaspora/blob/develop/lib/postzord/dispatcher.rb#L85
* https://github.com/diaspora/diaspora/blob/develop/lib/postzord/dispatcher.rb#L126
* https://github.com/diaspora/diaspora/blob/develop/lib/postzord/dispatcher.rb#L139
* https://github.com/diaspora/diaspora/blob/develop/app/workers/receive_local_batch.rb#L11-L12
* https://github.com/diaspora/diaspora/blob/develop/lib/postzord/receiver/local_batch.rb#L19
* https://github.com/diaspora/diaspora/blob/develop/lib/postzord/receiver/local_batch.rb#L36
* https://github.com/diaspora/diaspora/blob/develop/lib/diaspora/relayable.rb#L69 (the other codepath now, where we own the parent).

That should show all relevant "callback"/"hook" situations, the last part rather being for completeness.

It's probably fair if you just the testcase for `Federated::Generator` directly, using a test subclass that returns stub (`double`) values. Or you argue that it's opaque behaviour of the `CommentsController` or `CommentService` even, though that's a bit on the far edge of unit testing.